spinnaker/minimise.c

```diff
@@ -0,0 +1,154 @@
+/* Utility method to read a new copy of the routing table from SDRAM.        */
+void read_table(table_t *table, header_t *header)
+{
+  // Copy the size of the table
+  table->size = header->table_size;
+
+  // Allocate space for the routing table entries
+  table->entries = MALLOC(table->size * sizeof(entry_t));
+
+  // Copy in the routing table entries
+  spin1_memcpy((void *) table->entries,
+               (void *) header->entries,
+               sizeof(entry_t) * table->size);
+}
+/*****************************************************************************/
+
+/*****************************************************************************/
+/* Utility method to load a routing table to the router.                     */
+bool load_routing_table(table_t *table, uint32_t app_id)
+{
+  // Try to allocate sufficient room for the routing table.
+  uint32_t entry_id = rtr_alloc_id(table->size, app_id);
+
+  if (entry_id != 0)
+  {
+    // Load entries into the table (provided the allocation succeeded). Note
+    // that although the allocation included the specified application ID we
+    // also need to include it as the most significant byte in the route (see
+    // `sark_hw.c`).
+    for (uint32_t i = 0; i < table->size; i++)
+    {
+      entry_t entry = table->entries[i];
+      uint32_t route = entry.route | (app_id << 24);
+      rtr_mc_set(entry_id + i, entry.keymask.key, entry.keymask.mask, route);
+    }
+  }
+
+  // Indicate if we were able to allocate routing table entries.
+  return (entry_id != 0);
+};
+/*****************************************************************************/
+
+/*****************************************************************************/
+/* Utility method to sort routing table entries.                             */
+int compare_rte(const void *va, const void *vb)
+{
+  // Grab the keys and masks
+  keymask_t a = ((entry_t *) va)->keymask;
+  keymask_t b = ((entry_t *) vb)->keymask;
+
+  // Perform the comparison
+  return ((int) keymask_count_xs(a)) - ((int) keymask_count_xs(b));
+}
+/*****************************************************************************/
+
+/*****************************************************************************/
+void c_main(void)
+{
+  // Prepare to minimise the routing tables
+  header_t *header = (header_t *) sark_tag_ptr(1, 0);
+
+  // Load the routing table
+  table_t table;
+  read_table(&table, header);
+
+  // Store intermediate sizes for later reporting (if we fail to minimise)
+  uint32_t size_original, size_rde, size_mtrie, size_oc;
+  size_original = table.size;
+
+  // Try to load the table
+  if (!load_routing_table(&table, header->app_id))
+  {
+    // Otherwise remove default routes.
+    remove_default_routes_minimise(&table);
+    size_rde = table.size;
+
+    // Try to load the table
+    if (!load_routing_table(&table, header->app_id))
+    {
+      // Try to use m-Trie to minimise the routing table.
```

Comment:
It might be better to reload the table here and then remove default entries again after m-Trie has run.